src/audio/codec_adapter/codec_adapter.c

```diff
@@ -0,0 +1,528 @@
+{
+	comp_dbg(dev, "codec_adapter_params(): codec_adapter doesn't support .params() method.");
+	return 0;
+}
+
+static void
+codec_adapter_copy_from_source_to_lib(const struct audio_stream *source,
+				      const struct codec_processing_data *cpd,
+				      size_t bytes)
+{
+	/* head_size - available data until end of local buffer */
+	uint32_t head_size = MIN(bytes, audio_stream_bytes_without_wrap(source,
+				 source->r_ptr));
+	/* tail_size - residue data to be copied starting from the beginning
+	 * of the buffer
+	 */
+	uint32_t tail_size = bytes - head_size;
+	/* copy head_size to lib buffer */
+	memcpy_s(cpd->in_buff, cpd->in_buff_size, source->r_ptr, head_size);
+	if (tail_size)
+	/* copy reset of the samples after wrap-up */
+		memcpy_s((char *)cpd->in_buff + head_size, cpd->in_buff_size,
+			 audio_stream_wrap(source,
+					   (char *)source->r_ptr + head_size),
+					   tail_size);
+}
+
+static void
+codec_adapter_copy_from_lib_to_sink(const struct codec_processing_data *cpd,
+				    const struct audio_stream *sink,
+				    size_t bytes)
+{
+	/* head_size - free space until end of local buffer */
+	uint32_t head_size = MIN(bytes, audio_stream_bytes_without_wrap(sink,
+				 sink->w_ptr));
+	/* tail_size - rest of the bytes that needs to be written
+	 * starting from the beginning of the buffer
+	 */
+	uint32_t tail_size = bytes - head_size;
+
+	/* copy head_size to sink buffer */
+	memcpy_s(sink->w_ptr, sink->size, cpd->out_buff, head_size);
+	if (tail_size)
+	/* copy reset of the samples after wrap-up */
+		memcpy_s(audio_stream_wrap(sink,
+					   (char *)sink->w_ptr + head_size),
+			 sink->size,
+			 (char *)cpd->out_buff + head_size, tail_size);
+}
+
+static int codec_adapter_copy(struct comp_dev *dev)
+{
+	int ret = 0;
+	uint32_t bytes_to_process, processed = 0;
+	struct comp_data *cd = comp_get_drvdata(dev);
+	struct codec_data *codec = &cd->codec;
+	struct comp_buffer *source = cd->ca_source;
+	struct comp_buffer *sink = cd->ca_sink;
+	uint32_t codec_buff_size = codec->cpd.in_buff_size;
+	struct comp_copy_limits cl;
+
+	comp_get_copy_limits_with_lock(source, sink, &cl);
+	bytes_to_process = cl.frames * cl.source_frame_bytes;
+
+	comp_dbg(dev, "codec_adapter_copy() start: codec_buff_size: %d, sink free: %d source avail %d",
+		 codec_buff_size, sink->stream.free, source->stream.avail);
+
+	while (bytes_to_process) {
+		/* Proceed only if we have enough data to fill the lib buffer
+		 * completely. If you don't fill whole buffer
+		 * the lib won't process it.
+		 */
+		if (bytes_to_process < codec_buff_size) {
+			comp_dbg(dev, "codec_adapter_copy(): processed %d in this call %d bytes left for next period",
+				 processed, bytes_to_process);
+			break;
+		}
+
+		codec_adapter_copy_from_source_to_lib(&source->stream, &codec->cpd,
+						      codec_buff_size);
+		buffer_invalidate(source, codec_buff_size);
+		codec->cpd.avail = codec_buff_size;
+		//TODO: call codec specific processing here
```

Comment:
Anything todo here prior to merge ?

Reply:
No, this is added only to the initial commit in the series. Since you wanted it to be split in the file basics I had to add placeholders for the call which will occur later one. If you check the version of codec_adapter.c after the last commit you will see that `TODO` has been replaced with the actual call to codec processing function.

Reply:
ok, but for next time it's far easier and acceptable just to add a new file as a single commit (no need to show the development history here).

Reply:
oh, this should be C style comment too.